### PR TITLE
added angle reduction

### PR DESCRIPTION
**Pull request by nbronn**
Added transpiler pass that reduces angle to within -pi to pi range using modulo arithmetic, in order to make pulse attaching more efficient. Additionally added a unit test to test equivalences.

### CODE REVIEW

File: qiskit_research/utils/pulse_scaling.py [hunk 1]
```diff
@@ -94,7 +95,28 @@ def run(self, dag: DAGCircuit) -> DAGCircuit:
...
+
+class ReduceAngles(TransformationPass):
+    """Reduce angle of scaled pulses to between -pi and pi.
+
+    This works only after Parameters are bound. Gate strings
+    should only be single-parameter scaled pulses, i.e.
+    'rzx' and 'secr'.
+    """
+
+    def __init__(self, gate_strs: List[str]):
```
Comment (kevinsung) [inline qiskit_research/utils/pulse_scaling.py:109]: - rename `gate_strs -> gate_names`
- use `list[str] instead of List[str]`
- add a docstring where you documenting the argument. Say that it assumes each named gate has a single real-valued parameter.

File: qiskit_research/utils/pulse_scaling.py [hunk 2]
```diff
@@ -94,7 +95,28 @@ def run(self, dag: DAGCircuit) -> DAGCircuit:
...
+        super().__init__()
+        self._gate_strs = gate_strs
+
+    def run(self, dag: DAGCircuit) -> DAGCircuit:
+        for gate_str in self._gate_strs:
+            for grun in dag.collect_runs([gate_str]):
+                theta = grun[0].op.params[0]
+                grun[0].op.params[0] = (float(theta) + pi) % (2 * pi) - pi
```
Comment (kevinsung) [inline qiskit_research/utils/pulse_scaling.py:117]: Doesn't this only modify the first gate of the run? Also, please rename `grun` to simply `run`.

**Review by kevinsung — CHANGES_REQUESTED** [review]

File: test/utils/test_pulse_scaling.py [hunk 3]
```diff
@@ -138,3 +138,33 @@ def test_forced_rzz_template_match(self):
                 Operator(scale_qc_match.bind_parameters({theta: theta_set}))
             )
         )
+
+    def test_angle_reduction(self):
+        """Test Angle Reduction"""
```
Comment (kevinsung) [inline test/utils/test_pulse_scaling.py:143]: This test should take a different approach. It should directly instantiate a ReduceAngles pass, run the pass on a circuit with the gate, and check that the resultant gate has a reduced angle. That's all it needs to do.

**Review by kevinsung — CHANGES_REQUESTED** [review]

File: qiskit_research/utils/pulse_scaling.py (hunk 1 above)
Comment (nbronn) [reply qiskit_research/utils/pulse_scaling.py]: Done!

File: test/utils/test_pulse_scaling.py (hunk 3 above)
Comment (nbronn) [reply test/utils/test_pulse_scaling.py:143]: Rewrote pass to only check that angle is reduced.

File: test/utils/test_pulse_scaling.py [hunk 4]
```diff
@@ -138,3 +138,13 @@ def test_forced_rzz_template_match(self):
...
+        backend = FakeMumbai()
+
+        qc = QuantumCircuit(2)
+        qc.rzx(9 * np.pi / 2, 0, 1)
+        qc_s = scale_cr_pulses(qc, backend, param_bind={})
```
Comment (kevinsung) [inline test/utils/test_pulse_scaling.py:148]: I meant something like this, to directly test the pass:
```python
pass_manager = PassManager(ReduceAngles(["rzx"]))
qc_s = pass_manager.run(qc)
```

Comment (nbronn) [reply test/utils/test_pulse_scaling.py]: Ah yes I see, rewrote to just use the single pass.

File: test/utils/test_pulse_scaling.py [hunk 5]
```diff
@@ -138,3 +140,12 @@ def test_forced_rzz_template_match(self):
...
         )
+
+    def test_angle_reduction(self):
+        """Test Angle Reduction"""
+        qc = QuantumCircuit(2)
+        qc.rzx(9 * np.pi / 2, 0, 1)
+        pm = PassManager(ReduceAngles(["rzx"]))
+        qc_s = pm.run(qc)
+
+        self.assertAlmostEqual(qc_s.data[0].operation.params[0], np.pi / 2)
```
Comment (kevinsung) [inline test/utils/test_pulse_scaling.py:151]: One last thing: please test a few more values, including ones that should reduce to negative values, and edge cases like `-pi`

Comment (nbronn) [reply test/utils/test_pulse_scaling.py]: Added a couple more checks.

**Review by kevinsung — APPROVED** [review]
LGTM
